### ISSUE TITLE
How can cleanlab deal with too many labels?

### ISSUE THREAD

**Issue by Acquaintancy-xmh** [cgnorthcutt/cleanlab#34]
HI, I'm using cleanlab to prune noise labels in my dataset. However, I find out that cleanlab was ineffective. There are 3900 classes in my dataset with almost 30million examples. The psx matrix is sparse. What can I do with this situation?

**Comment by cgnorthcutt** [comment]
Hi. What is your cross validation accuracy on your noisy train set with the model you're using? 

Also, did you obtain psx using cross validation?

**Comment by Acquaintancy-xmh** [comment]
My cross validation is 77% , with almost 30% noisy labels in my  dataset.
The psx matrix did generated from cross validation (5 folds), which each fold contains 7.2million. However, since the dataset was too large,  I randomly selected 1,000k examples for psx.

**Comment by cgnorthcutt** [comment]
Got it. 77% accuracy with 30% noisy labels is surprising. Typically with X% error you would expect 100 - X % accuracy. So your model may be over-fitting to the noise, so that it predicts the noisy labels, not the true labels, which will make it hard for your model to generalize to true labels. However, with 77% accuracy on noisy labels, cleanlab should work for you. What is the accuracy on clean labels?

I did not understand your last sentence -- what is the shape of your psx -- is it 1000 x 3900? Are you training on only 1000? you can fit psx in memory more easily by using float16 (or even uint8 if you're thoughtful about it).

**Comment by Acquaintancy-xmh** [comment]
Sorry for my mistake. 77% accuracy was on clean labels which were picked manually. I was training on 4 folds, however, because of the memory limit, I picked 1000k examples from the other fold (7.2million). The psx is 1000k x 3900.
I observed my psx was sparse .Since there was too many classes, for some examples, the p of some class is less than 1e-5. I wonder that how can i fix this problem. Should I test on a bigger test set than 1000k?

**Comment by cgnorthcutt** [comment]
When you say 1000k, do you mean 1000 or 1000 x 1000 = 1,000,000?

**Comment by Acquaintancy-xmh** [comment]
1000 * 1000 = 1,000,000

**Comment by cgnorthcutt** [comment]
Are you using datatype float16 for psx? Do not use float64

**Comment by Acquaintancy-xmh** [comment]
I'm using float64. I will try float 16. Thanks for your patient.

**Comment by cgnorthcutt** [comment]
If you used 10 fold training, then each prediction would only be 3mil examples, and you'd have 10 psx for each fold of shape 3mil x 3900, and each one will be float16, which takes up 17.4 GB in memory, for each of the ten folds. Is that possible for you?

**Comment by cgnorthcutt** [comment]
With four fold, it would be 7.5 x 3900 x 2 bytes (16 bits for float 16), which is 43.5 GB per fold. Definitely getting up there in size, but if you're able to train on a dataset with 30 million examples, hopefully you can handle that.

**Comment by cgnorthcutt** [comment]
Last thing - if you want to take advantage of sparse matrices, you just need to convert the psx to scipy csr matrices. cleanlab should be compatible with these. See: https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.sparse.csr_matrix.html

No promises here (use at your own risk), but this has worked for a number of others.

**Comment by Acquaintancy-xmh** [comment]
Thanks for your suggestions. I'll try the scipy csr matrices. I hope my gpu could handle a dataset with 30million examples.

**Comment by Acquaintancy-xmh** [comment]
One more question, will cleanlab perform better on large psx , or does it have nothing to do with psx size?

**Comment by cgnorthcutt** [comment]
If you don't give cleanlab the complete psx, it cannot accurately estimate the noise thresholds as it will not see a significant portion of the noise in your data. There are some situations where this is okay (like randomly uniform noise), but those situations are not realistic. In the real world, you need to give cleanlab the full psx matrix so it can find the noise thresholds, find all the error in your dataset, and provide clean data for training.

**Comment by Acquaintancy-xmh** [comment]
Thanks. That did help me a lot.
